### ISSUE TITLE
Mypy can not check type of object

### ISSUE THREAD

**Issue by futorio** [python/mypy#15034]
**Bug Report**

Hi! I have a function which check that given object (obj) has given type (target_type).

```python
T = TypeVar("T")


def check_type(obj: T, target_type: Type[T]) -> T:
    if not isinstance(obj, target_type):
        raise TypeError(obj, target_type)

    return obj
```

So i expected that mypy find an error in code like `check_type(123, str)`, but this code passed the test.

**To Reproduce**

[mypy playground](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict%2Ccheck-untyped-defs&gist=5e4d87bd60e2d016c5d9269a6d344dea)

**Expected Behavior**

Failed status with message like: "main.py:14: error: Argument 1 to "check_type" has incompatible type "int"; expected "Type[str]"  [arg-type]"

**Actual Behavior**

Success: no issues found in 1 source file

**Your Environment**

- Mypy version used: 1.2.0
- Mypy command-line flags: --strict --check-untyped-defs
- Python version used: 3.10

**Comment by thq919** [comment]
very same issue, needs to be solved

**Comment by tmke8** [comment]
Mypy is inferring `T` to be `object`. And `Type[str]` is compatible with `Type[object]` because `Type[]` is [covariant](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types):

```python
c: Type[object] = str  # OK
```

For your use case, you could use `typing.assert_type` instead (though that sadly has no runtime check).
